test/object-store/realm.cpp

```diff
@@ -1206,6 +1207,104 @@ TEST_CASE("Get Realm using Async Open", "[sync][pbs][async open]") {
+                REQUIRE_FALSE(error.is_fatal);
+                return;
+            }
+            if (error.status.code() == ErrorCodes::AuthError) {
+                REQUIRE(error.is_fatal);
+            }
+        };
+
+        auto valid_token = user->access_token();
+        // User should be logged in at this point
+        bool not_authorized = false;
+        bool token_refresh_called = false;
+        bool location_refresh_called = false;
+
+        TestMode mode = GENERATE(location_fails, token_fails, token_not_authorized);
+
+        SECTION("access token expired when realm is opened") {
+            logger->trace(">>> access token expired at start - mode: %1", mode);
+            user->update_access_token(std::move(expired_token));
+        }
+        SECTION("access token expired when websocket connects") {
+            logger->trace(">>> access token expired by websocket - mode: %1", mode);
+            not_authorized = true;
+        }
+        SECTION("access token expired when websocket connects") {
+            logger->trace(">>> websocket returns connection failed - mode: %1", mode);
+        }
+
+        transport->request_hook = [&](const app::Request& req) -> std::optional<app::Response> {
+            std::lock_guard<std::mutex> lock(mutex);
+            if (req.url.find("/auth/session") != std::string::npos) {
+                token_refresh_called = true;
+                if (mode == token_not_authorized) {
+                    return app::Response{403, 0, {}, "403 not authorized"};
+                }
+                if (mode == token_fails) {
+                    return app::Response{0, 28, {}, "Operation timed out"};
+                }
+            }
+            else if (req.url.find("/location") != std::string::npos) {
+                location_refresh_called = true;
+                if (mode == location_fails) {
+                    // Fake "offline/request timed out" custom error response
+                    return app::Response{0, 28, {}, "Operation timed out"};
+                }
+            }
+            return std::nullopt;
+        };
+
+        socket_provider->websocket_connect_func = [&]() -> std::optional<SocketProviderError> {
+            if (not_authorized) {
+                not_authorized = false; // one shot
+                return SocketProviderError(sync::websocket::WebSocketError::websocket_unauthorized,
+                                           "403 not authorized");
+            }
+            return SocketProviderError(sync::websocket::WebSocketError::websocket_connection_failed,
+                                       "Operation timed out");
+        };
+
+        auto task = Realm::get_synchronized_realm(config);
+        auto [promise, async_future] = util::make_promise_future<void>();
+        task->start(
+            [async_promise = util::CopyablePromiseHolder<void>(std::move(promise))](auto ref, auto error) mutable {
```

Comment:
This can capture `promise` by reference rather than wrapping it in CopyablePromiseHolder since it's synchronously waiting on the future before the promise would go out of scope.

I think `util::make_promise_future<std::exception_ptr>()` is more straightforward here; the test is just validating that a non-null exception_ptr is being produced so it can use that as the value directly rather than converting it to an error-case Status.